File: js-node/day7/src/filesystem.js
import { open } from "node:fs/promises";

export class FileSystem {
  /** @type {FileSystemEntry} */
  root;
  /** @type {FileSystemEntry} */
  currentWorkingDirectory;

  constructor() {
    this.root = FileSystemEntry.newDirectory("/");
  }

  /**
   * @param {string} terminalOutputFile
   */
  static async build(terminalOutputFile) {
    const input = await open(terminalOutputFile);
    const filesystem = new FileSystem();

    for await (const line of input.readLines()) {
      if (line.startsWith("$")) {
        // Handle commands
        if (line.startsWith("$ cd")) {
          filesystem.changeDirectory(line.substring(5));
        }
      } else {
        // Create file and add entry
        if (line.startsWith("dir")) {
          const directoryName = line.substring(4);

          filesystem.makeDirectory(directoryName);
        } else {
          const [fileSize, fileName] = line.split(" ");

          filesystem.touch(fileName, Number(fileSize));
        }
      }
    }

    await input.close();

    return filesystem;
  }

  /**
   * @param {"/" | ".." | string} directory
   */
  changeDirectory(directory) {
    if (directory === "/") {
      this.currentWorkingDirectory = this.root;
    } else if (directory === "..") {
      this.currentWorkingDirectory = this.currentWorkingDirectory.parent;
    } else {
      const childDirectory = this.currentWorkingDirectory.directoryEntries.find(
        (entry) => entry.name === directory,
      );

      if (!childDirectory) {
        throw new Error(`Cannot change to directory ${directory}`);
      }

      this.currentWorkingDirectory = childDirectory;
    }
  }

  /**
   * @param {string} name
   */
  makeDirectory(name) {
    const entry = FileSystemEntry.newDirectory(name);

    entry.parent = this.currentWorkingDirectory;
    this.currentWorkingDirectory.addEntry(entry);
  }

  /**
   * @param {string} name
   * @param {number} size
   */
  touch(name, size) {
    const entry = FileSystemEntry.newFile(name, size);

    entry.parent = this.currentWorkingDirectory;
    this.currentWorkingDirectory.addEntry(entry);
  }

  /**
   * @template T
   * @param {(aggregatedValue: T, currentDirectory: FileSystemEntry) => T} aggregateFn
   * @param {T} initialValue
   */
  traverseFilesystem(aggregateFn, initialValue) {
    const directoriesToCheck = [this.root];
    let aggregatedValue = initialValue;

    while (directoriesToCheck.length > 0) {
      const currentDir = directoriesToCheck.pop();

      aggregatedValue = aggregateFn(aggregatedValue, currentDir);

      for (const child of currentDir.directoryEntries) {
        if (child.isDirectory) {
          directoriesToCheck.push(child);
        }
      }
    }

    return aggregatedValue;
  }
}

export class FileSystemEntry {
  /** @type {string} */
  name;
  /** @type {boolean} */
  isDirectory;
  /** @type {number} */
  #size = 0;
  /** @type {FileSystemEntry[]} */
  directoryEntries = [];
  /** @type {FileSystemEntry | undefined} */
  parent;

  get size() {
    if (this.isDirectory) {
      return this.directoryEntries.reduce((sum, entry) => sum + entry.size, 0);
    } else {
      return this.#size;
    }
  }

  /**
   * @param {string} name
   * @param {number} size
   */
  static newFile(name, size) {
    const entry = new FileSystemEntry();

    entry.name = name;
    entry.#size = size;
    entry.isDirectory = false;

    return entry;
  }

  /**
   * @param {string} name
   */
  static newDirectory(name) {
    const entry = new FileSystemEntry();

    entry.name = name;
    entry.isDirectory = true;

    return entry;
  }

  /**
   * @param {FileSystemEntry} entry
   */
  addEntry(entry) {
    if (this.isDirectory) {
      this.directoryEntries.push(entry);
    } else {
      throw new Error("Cannot add this entry to a file.");
    }
  }
}

File: rust/day07/src/filesystem.rs
use std::{io::BufRead, rc::Rc};

use crate::filesystem_entry::FilesystemEntry;

pub struct Filesystem {
    root: Rc<FilesystemEntry>,
    current_working_directory: Rc<FilesystemEntry>,
}

impl Filesystem {
    pub fn new() -> Self {
        let root = Rc::new(FilesystemEntry::new_directory(None));

        Self {
            root: Rc::clone(&root),
            current_working_directory: root, //RefCell::new(root),
        }
    }

    pub fn build_with_descriptor(&mut self, descriptor: &mut dyn BufRead) {
        for line in descriptor.lines().map(|line| line.unwrap()) {
            if line.starts_with("$") {
                // Handle command
                if line.starts_with("$ cd") {
                    self.change_directory(line.strip_prefix("$ cd ").unwrap());
                }
            } else {
                // Add entry
                if line.starts_with("dir") {
                    let name = line.strip_prefix("dir ").map(String::from).unwrap();

                    self.make_directory(name);
                } else {
                    let mut line = line.split_whitespace();
                    let size: u32 = line.next().and_then(|size| size.parse().ok()).unwrap();
                    let name = line.next().map(String::from).unwrap();

                    self.touch(name, size);
                }
            }
        }
    }

    pub fn root(&self) -> &Rc<FilesystemEntry> {
        &self.root
    }

    pub fn size(&self) -> u32 {
        self.root.size()
    }

    fn change_directory(&mut self, directory: &str) {
        let new_directory = match directory {
            "/" => Some(Rc::clone(&self.root)),
            ".." => self.current_working_directory.parent(),
            child => self.current_working_directory.find_child(child),
        };

        new_directory.map(|dir| self.current_working_directory = dir);
    }

    fn make_directory(&self, name: String) {
        if let FilesystemEntry::Directory { entries, .. } = self.current_working_directory.as_ref()
        {
            let cwd = Rc::clone(&self.current_working_directory);
            let new_directory = FilesystemEntry::new_directory(Some(cwd));

            entries.borrow_mut().insert(name, Rc::new(new_directory));
        }
    }

    fn touch(&self, name: String, size: u32) {
        if let FilesystemEntry::Directory { entries, .. } = self.current_working_directory.as_ref()
        {
            let cwd = Rc::clone(&self.current_working_directory);
            let new_file = FilesystemEntry::new_file(size, Some(cwd));

            entries.borrow_mut().insert(name, Rc::new(new_file));
        }
    }
}
